### PR TITLE
[FEATURE] Empty bins in layout

### PR DESCRIPTION
**Pull request by eseiler**

### CODE REVIEW

**Comment by seqan-actions** [comment]
Documentation preview available at https://docs.seqan.de/preview/seqan/hibf/251

File: include/hibf/layout/hierarchical_binning.hpp [hunk 1]
```diff
@@ -68,7 +69,8 @@ class hierarchical_binning
         config{config_},
         data{std::addressof(data_)},
         num_user_bins{data->positions.size()},
-        num_technical_bins{data->previous.empty() ? config.tmax : needed_technical_bins(num_user_bins)}
+        num_technical_bins{data->previous.empty() ? tmax_without_empty_bins(config.tmax, config.empty_bin_fraction)
+                                                  : needed_technical_bins(num_user_bins)}
     {
         assert(data != nullptr);
```
Comment (eseiler) [inline include/hibf/layout/hierarchical_binning.hpp:75]: This line could be removed, no?
data_ is a reference, so this pointer can never be null

File: src/layout/hierarchical_binning.cpp [hunk 2]
```diff
@@ -79,7 +80,8 @@ size_t hierarchical_binning::execute()
 
 [[nodiscard]] size_t hierarchical_binning::needed_technical_bins(size_t const requested_num_ub) const
 {
-    return std::min<size_t>(next_multiple_of_64(requested_num_ub), config.tmax);
+    size_t const needed = std::min<size_t>(next_multiple_of_64(requested_num_ub), config.tmax);
+    return tmax_without_empty_bins(needed, config.empty_bin_fraction);
```
Comment (smehringer) [inline src/layout/hierarchical_binning.cpp:84]:
```suggestion
    return tmax_without_empty_bins(needed, config.empty_bin_fraction); // subtract empty bins before layouting
```

File: src/layout/hierarchical_binning.cpp [hunk 3]
```diff
@@ -406,16 +408,21 @@ void hierarchical_binning::update_libf_data(data_store & libf_data, size_t const
...
         // recursively call hierarchical binning if there are still too many UBs
         return hierarchical_binning{libf_data, config}.execute(); // return id of maximum technical bin
     }
     else
     {
         // use simple binning to distribute remaining UBs
-        return simple_binning{libf_data, 0}.execute(); // return id of maximum technical bin
+        // Simple binning is not bound by config.tmax
+        size_t const num_user_bins = next_multiple_of_64(number_of_user_bins);
+        size_t const number_of_technical_bins = tmax_without_empty_bins(num_user_bins, config.empty_bin_fraction);
```
Comment (smehringer) [inline src/layout/hierarchical_binning.cpp:424]:
```suggestion
        size_t const number_of_technical_bins = tmax_without_empty_bins(num_user_bins, config.empty_bin_fraction); // subtracts empty bins
```

File: test/unit/hibf/layout/hierarchical_binning_test.cpp [hunk 4]
```diff
@@ -68,6 +68,81 @@ TEST(hierarchical_binning_test, small_example)
     EXPECT_RANGE_EQ(hibf_layout.user_bins, expected_user_bins);
 }
 
+TEST(hierarchical_binning_test, small_example_two)
```
Comment (smehringer) [inline test/unit/hibf/layout/hierarchical_binning_test.cpp:71]: instead of introducing `small_example_two` you could directly add an empty bin test that has tmax = 5, an empty bin fraction that results in layouting with 4 technical bins, and then the result should be equivalent to the result of `small_example`

**Review by smehringer — CHANGES_REQUESTED** [review]

File: include/hibf/misc/tmax_without_empty_bins.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace seqan::hibf
+{
+
+/*!\brief Returns the number of technical bins available for use.
+ * \param[in] tmax     The total number of bins.
+ * \param[in] fraction The fraction of the total number of bins that should be empty.
+ * \ingroup hibf
+ * \sa https://godbolt.org/z/cMjbM39vj
+ */
+[[nodiscard]] constexpr size_t tmax_without_empty_bins(size_t const tmax, double const fraction) noexcept
```
Comment (smehringer) [inline include/hibf/misc/tmax_without_empty_bins.hpp:22]:
```suggestion
[[nodiscard]] constexpr size_t subtract_empty_bins(size_t const tmax, double const fraction) noexcept
```

File: include/hibf/config.hpp [hunk 6]
```diff
@@ -230,6 +231,9 @@ struct config
      */
     size_t tmax{};
 
+    //!\brief The percentage of empty bins in the layout.
+    double empty_bin_fraction{};
```
Comment (smehringer) [inline include/hibf/config.hpp:235]: please write a more detailed description

File: include/hibf/layout/hierarchical_binning.hpp [hunk 7]
```diff
@@ -68,10 +69,9 @@ class hierarchical_binning
         config{config_},
         data{std::addressof(data_)},
         num_user_bins{data->positions.size()},
-        num_technical_bins{data->previous.empty() ? config.tmax : needed_technical_bins(num_user_bins)}
-    {
-        assert(data != nullptr);
-    }
+        num_technical_bins{data->previous.empty() ? subtract_empty_bins(config.tmax, config.empty_bin_fraction)
```
Comment (SGSSGene) [inline include/hibf/layout/hierarchical_binning.hpp:72]: Maybe this variable should be renamed? something like "num_technical_bins_in_use" or "occupied_technical_bins". So it gets clear that it does not include the empty bin?

**Review by SGSSGene — APPROVED** [review]

File: include/hibf/layout/hierarchical_binning.hpp (hunk 7 above)
Comment (eseiler) [reply include/hibf/layout/hierarchical_binning.hpp:72]: The layout isn't aware how many technical bins there actually are.
In the context of the layout, there are only ever `num_technical_bins`.
The algorithm doesn't really care whether you want 128 or 125 technical bins.

The meaning of occupied/unoccupied is introduced outside the layout.
The HIBF knows that it wants empty bins and has a tmax of 128.
So it then requests a layout for 125 bins. And the last 3 are empty.

Does that make sense?

Comment (SGSSGene) [reply include/hibf/layout/hierarchical_binning.hpp:72]: So in your example, it is specifically leaving 3 bins out of the layout computation so it can be filled later, as it likes.

**Review by SGSSGene — APPROVED** [review]
